### PR TITLE
Partial fix for LoftQ + int8 quantization

### PR DESCRIPTION
**Pull request by githubnemo**
With transformers v5 the GPU examples started failing for LoftQ, mainly for 8bit quantization. This is also reported in PR #3021. The main reason for this is that with transformers v5 the dtype of the layers is consistent across the model whereas before some layers were in float16 and some in float32. This masked wrong behavior.

It turns out that LoftQ always assumed that if `n_bits==8` the weights are in fp8. In reality, at least for bitsandbytes, this is not true. bnb's `load_in_8bit` uses int8. This PR changes the quantization residual computation to be compatible with int8 if `n_bits==8`, so that at least the computed residual for the loftq initialization are correct.

This alone does not mean that LoftQ tests for 8bit are passing. They assume that the MSE of LoftQ 'corrected' models compared to an unquantized baseline model are at least 3% truer than the logits of the quantized reference model. For int8 this is not guaranteed since `bnb.nn.Linear8bitLt` not only quantizes the weights but it also quantizes the inputs in it's `.forward()`. The resulting quantization error cannot be compensated by the static LoftQ initialization. Since there is technically a way to compensate for this error, I've kept the LoftQ initialization for int8, added an example and changed the 8 bit tests to only assume rough parity with the quantized reference model.

The example (`./examples/loftq_finetuning/int8_correction.py`) that show-cases how to recover the activation quantization error can also be run with `--no-loftq` which skips the LoftQ initialization and uses the LoRA no-op initialization instead and the example still recovers the majority of the error (e.g. 90% improvement instead of 91% over quantized reference).

Therefore there are the following open tasks:

- discuss removal of int8 from LoftQ completely (since it can be argued that it doesn't seem to have such a big impact but this needs review and possibly better testing)
- discuss the now failing (marked as xfail) tests for nf4 for t5 when loftq_iter > 1. I don't have an intuition what's happening here but the error seems to roughly double per iteration.

### CODE REVIEW

File: examples/loftq_finetuning/int8_correction.py [hunk 1]
```diff
@@ -0,0 +1,250 @@
+"""
+Script to show-case how to offset quantization error with LoRA / LoftQ
+when dealing with quantizations that both quantize weights and activations.
+This is the case for bnb int8, for example but also other quantizations
```
Comment (BenjaminBossan) [inline examples/loftq_finetuning/int8_correction.py]:
```suggestion
This is the case for bnb int8, for example, but also other quantizations
```

File: examples/loftq_finetuning/int8_correction.py [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+    return torch.pow(a - b, 2).mean().item()
+
+
+def get_model(*args, **kwargs):
+    if task_type == TaskType.CAUSAL_LM:
+        return AutoModelForCausalLM.from_pretrained(*args, **kwargs)
+    return AutoModelForSeq2SeqLM.from_pretrained(*args, **kwargs)
+
+
+tokenizer = AutoTokenizer.from_pretrained(model_id)
```
Comment (BenjaminBossan) [inline examples/loftq_finetuning/int8_correction.py:231]: Just a suggestion to move this all into a `main` function to avoid computing all this if only `--help` is called.

File: examples/loftq_finetuning/int8_correction.py [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+loftq_logits = get_logits(loftq_model, inputs)
+
+mse_loftq = mse(ref_logits, loftq_logits)
+mse_qref = mse(ref_logits, qref_logits)
+
+
+print(f"{model_id=}{device=}")
+print(f"{mse_qref=}, {mse_loftq=}")
+assert mse_loftq < (mse_qref / 1.05), f"{mse_loftq} >= {mse_qref / 1.05}"
+print(f"relative improvement: {(mse_qref - mse_loftq) / mse_qref * 100:.2f}%")
```
Comment (BenjaminBossan) [inline examples/loftq_finetuning/int8_correction.py]: IMO, "improvement" is a bit ambiguous here, how about "reduction of error through LoftQ"?

File: examples/loftq_finetuning/int8_correction.py [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+The quantization error for the affected values is much lower and not
+static anymore, LoftQ is not able to deal with this. While technically
+possible, this script doesn't filter out these masked values and it
+is probably not worth the effort.
+
+Note: Some rudimentary testing showed that the LoftQ mitigation is still
+more effective than tuning threshold values but YMMV.
+
+Note: LoftQ is not doing the heavy lifting in this script's case. The error
+between LoftQ and zeroed LoRA is only about two percent points. This effect is
```
Comment (BenjaminBossan) [inline examples/loftq_finetuning/int8_correction.py]: How did you determine this 2%?

File: src/peft/utils/loftq_utils.py [hunk 5]
```diff
@@ -195,8 +204,8 @@ def loftq_init(weight: Union[torch.Tensor, torch.nn.Parameter], num_bits: int, r
     else:
         raise ValueError("bitsandbytes is not available, please install it to use LoftQ.")
 
-    if num_bits not in [2, 4, 8]:
-        raise ValueError("Only support 2, 4, 8 bits quantization")
+    if num_bits not in [4, 8]:
+        raise ValueError("Only support nf4 and int8 quantization is supported")
```
Comment (BenjaminBossan) [inline src/peft/utils/loftq_utils.py]:
```suggestion
        raise ValueError("Only nf4 and int8 quantization is supported")
```

File: tests/test_gpu_examples.py [hunk 6]
```diff
@@ -2979,12 +2980,13 @@ class TestLoftQ:
...
-    def get_error_factor(self, device):
+    OPT_MODEL_ID = "facebook/opt-125m"
+
+    def get_error_factor(self, device, n_bits):
         # The error factor indicates by how much the quantization error should be decreased when using LoftQ compared to
         # quantization without LoftQ. Thus 1.03 means that the error should be decreased by 3% at least. This is a very
         # conservative value to prevent flakiness, in practice most gains are > 1.5
-        error_factor = 1.005 if device in ("xpu", "cpu") else 1.03
-        return error_factor
+        return 1.05 if device in ("xpu", "cpu") else 1.05
```
Comment (BenjaminBossan) [inline tests/test_gpu_examples.py]:
```suggestion
        return 1.05
```

File: tests/test_gpu_examples.py [hunk 7]
```diff
@@ -3087,67 +3108,31 @@ def get_errors(
         mse_quantized = torch.pow(logits_base - logits_quantized, 2).mean()
         mae_loftq = torch.abs(logits_base - logits_loftq).mean()
         mse_loftq = torch.pow(logits_base - logits_loftq, 2).mean()
+
+        print(f"{bits=} {loftq_iter=} {device=} {model_id=}")
```
Comment (BenjaminBossan) [inline tests/test_gpu_examples.py]: Remove?

File: tests/test_other.py [hunk 8]
```diff
@@ -633,3 +633,11 @@ def test_get_modules_tied_to_embedding_peft(self, model_id, tied_weights_type):
...
 
             assert expected == modules
+
+
+class TestLoftQDeprecation:
+    def test_nfquantizer_deprecation(self):
+        from peft.utils.loftq_utils import NFQuantizer
+
+        with pytest.warns(match="NFQuantizer is deprecated") as record:
+            _ = NFQuantizer()
```
Comment (BenjaminBossan) [inline tests/test_other.py]:
```suggestion
            _ = NFQuantizer(device="cpu")
```

I think this should fix the CI error.

File: tests/test_other.py [hunk 9]
```diff
@@ -633,3 +633,11 @@ def test_get_modules_tied_to_embedding_peft(self, model_id, tied_weights_type):
             modules = peft_model._get_module_names_tied_with_embedding()
 
             assert expected == modules
+
+
+class TestLoftQDeprecation:
```
Comment (BenjaminBossan) [inline tests/test_other.py:639]: Let's add a TODO to remove this in 0.20.0

File: tests/test_gpu_examples.py [hunk 10]
```diff
@@ -3156,49 +3141,56 @@ def test_bloomz_loftq_8bit_iter_5(self, device, tmp_path):
...
+        if n_bits == 4:
+            # next, check that LoftQ quantization errors are smaller than LoRA errors by a certain margin
+            error_factor = self.get_error_factor(device, n_bits)
+            assert mse_loftq < (mse_quantized / error_factor)
+            assert mae_loftq < (mae_quantized / error_factor)
+        elif n_bits == 8:
+            # for int8 compensation we're happy to achieve parity but without additional measures we will
+            # not be better since int8 also quantizes the layer input which introduces error loftq can't
+            # compensate
+            assert torch.allclose(mse_loftq, mse_quantized, atol=0.03)
```
Comment (BenjaminBossan) [inline tests/test_gpu_examples.py]: Just to be sure: The error may even increase in this case?

File: examples/loftq_finetuning/int8_correction.py (hunk 2 above)
Comment (githubnemo) [reply examples/loftq_finetuning/int8_correction.py:231]: In that case `parse_args()` would call `sys.exit(1)` anyway right? So, no computation would be done.

File: tests/test_gpu_examples.py (hunk 10 above)
Comment (githubnemo) [reply tests/test_gpu_examples.py]: Yes. I think it is a bit dependent on the number of input sequences, it seems that the variance is quite high. Maybe we should include more than one test sequence.

**Comment by githubnemo** [comment]
@BenjaminBossan I added multiple input sequences to the tests to reduce the variance a bit. Unfortunately it turns out that we have to increase the variance (and the rank/alpha as it seems) a bit to have a meaningful effect. I'm a bit surprised that 4bit quantization was affected as well (at least a few tests failed with r=8).

I've also added padding masking in the MSE computation which increased some LoftQ scores but didn't change the overall results and margins too much.
